### PR TITLE
shared-module/usb_hid: allow HID to wake sleeping host computer

### LINKED ISSUE #5380: How to enable keyboard to wake sleeping computer?
Hi,

I'm trying to figure out how to make an emulated USB keyboard (using a neokey 1x4) wake the computer it's attached to.  I've verified the QT PY RP2040 is getting power and behaves normally when the computer is on, but it seems to go haywire when the pc is asleep and i press a key (windows 10).

Is this possible to support somehow with the current libraries?

### LINKED ISSUE #8774: wakeup feature for usb_hid package
usb_hid packages support keyboard, mouse, consumer_contol features based on tinyusb. The examples of tinyusb can support wakeup in host sleep mode but Circuitpython usb_hid packages has no code to support it. I added some code from tinyusb example to Circuitpython usb implementation code. I tested this modification on esp32s2 board and works fine as I expected.

Here are the diff file.

<pre><code>
diff --git a/shared-module/usb_hid/Device.c b/shared-module/usb_hid/Device.c
index b28de9c1b1..7d0e6eff3a 100644
--- a/shared-module/usb_hid/Device.c
+++ b/shared-module/usb_hid/Device.c
@@ -229,12 +229,16 @@ void common_hal_usb_hid_device_send_report(usb_hid_device_obj_t *self, uint8_t *
         RUN_BACKGROUND_TASKS;
     }

-    if (!tud_hid_ready()) {
-        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("USB busy"));
-    }
+    if (!tud_suspended()) {
+        if (!tud_hid_ready()) {
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("USB busy"));
+        }

-    if (!tud_hid_report(report_id, report, len)) {
-        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("USB error"));
+        if (!tud_hid_report(report_id, report, len)) {
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("USB error"));
+        }
+    } else {
+        tud_remote_wakeup();
     }
 }

diff --git a/supervisor/shared/usb/usb_desc.c b/supervisor/shared/usb/usb_desc.c
index 811f61724f..9f932ddbfe 100644
--- a/supervisor/shared/usb/usb_desc.c
+++ b/supervisor/shared/usb/usb_desc.c
@@ -105,7 +105,7 @@ static const uint8_t configuration_descriptor_template[] = {
 #define CONFIG_NUM_INTERFACES_INDEX (4)
     0x01,        // 5 bConfigurationValue
     0x00,        // 6 iConfiguration (String Index)
-    0x80,        // 7 bmAttributes
+    0x80 | TUSB_DESC_CONFIG_ATT_REMOTE_WAKEUP,        // 7 bmAttributes
     0x32,        // 8 bMaxPower 100mA
 };
</code></pre>

### PR DESCRIPTION
**Pull request by dhalbert**
- Fixes #8774.
- Fixes #5380.

Thanks very much to @meesokim for code to enable `usb_hid` to wake a sleeping host computer, and to @hathach for implementing this ability in TinyUSB in the past.

Tested successfully with MacroPad RP2040, Feather nRF52840, and Circuit Playground Express on Windows and macOS.

On Linux not all HID devices are automatically enabled to do wakeup. See @hathach's [comment](https://github.com/adafruit/circuitpython/issues/5380#issuecomment-925044500). See also https://askubuntu.com/questions/848698/wake-up-from-suspend-using-wireless-usb-keyboard-or-mouse-for-any-linux-distro/874701 for details on how to change these settings. I would have thought there was an easier way, and maybe there is for some Linux distributions.

For example on my Linux box, by default only my primary keyboard is enabled for wakeup. This is a nuisance and will mean that people using, say, a MacroPad on Linux will need to adjust these settings if they want the MacroPad to be able to initiate wakeup.
```sh
$  grep . /sys/bus/usb/devices/*/product
/sys/bus/usb/devices/1-12/product:HD Pro Webcam C920
/sys/bus/usb/devices/1-1/product:Freestyle Edge RGB Keyboard
/sys/bus/usb/devices/1-2.5/product:USB Billboard Device   
/sys/bus/usb/devices/1-2/product:USB2.0 Hub             
/sys/bus/usb/devices/1-3/product:Microsoft Basic Optical Mouse 
/sys/bus/usb/devices/1-6/product:Macropad RP2040
/sys/bus/usb/devices/2-2.1/product:USB 10/100/1000 LAN
/sys/bus/usb/devices/2-2/product:USB3.0 Hub             
/sys/bus/usb/devices/usb1/product:xHCI Host Controller
/sys/bus/usb/devices/usb2/product:xHCI Host Controller

$  grep . /sys/bus/usb/devices/*/power/wakeup
/sys/bus/usb/devices/1-10/power/wakeup:disabled
/sys/bus/usb/devices/1-1/power/wakeup:enabled
/sys/bus/usb/devices/1-2.5/power/wakeup:disabled
/sys/bus/usb/devices/1-2/power/wakeup:disabled
/sys/bus/usb/devices/1-3/power/wakeup:disabled
/sys/bus/usb/devices/1-6/power/wakeup:disabled
/sys/bus/usb/devices/1-8/power/wakeup:disabled
/sys/bus/usb/devices/2-2.1/power/wakeup:disabled
/sys/bus/usb/devices/2-2/power/wakeup:disabled
/sys/bus/usb/devices/usb1/power/wakeup:disabled
/sys/bus/usb/devices/usb2/power/wakeup:disabled
```

### CODE REVIEW

**Comment by dhalbert** [comment]
A few `ru` builds don't fit anymore. @tannewt's upcoming error message shrinking should fix this, so we can ignore those failures.

File: shared-module/usb_hid/Device.c [hunk 1]
```diff
@@ -229,12 +229,16 @@ void common_hal_usb_hid_device_send_report(usb_hid_device_obj_t *self, uint8_t *
...
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("USB busy"));
+        }
 
-    if (!tud_hid_report(report_id, report, len)) {
-        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("USB error"));
+        if (!tud_hid_report(report_id, report, len)) {
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("USB error"));
+        }
+    } else {
+        tud_remote_wakeup();
```
Comment (tannewt) [inline shared-module/usb_hid/Device.c:241]: We should probably change the function description to note that wakeup drops a report.

**Review by tannewt — CHANGES_REQUESTED** [review]
One comment request. Thanks for doing this!

**Review by tannewt — APPROVED** [review]
Looks good! You'll want to merge to get the file size reductions.

**Comment by bendavis78** [comment]
@dhalbert Is there any documentation on how to use this new wakeup feature?

**Comment by dhalbert** [comment]
> @dhalbert Is there any documentation on how to use this new wakeup feature in python?

It should "just work". If the HID device sends a report (e.g. a key press) to the host, the host should wake up. This is how a keypress on your regular keyboard works.

**Comment by bendavis78** [comment]
I see, thanks! I tested on a Mac and a keypress does in fact wake it up. Side-note: I'm using a Pico W for this project. 

My current problem is with waking a Linux host. I've tried the solution explained in the SO post:

```
$ lsusb | grep Adafruit
Bus 003 Device 018: ID 239a:8120 Adafruit Pico W
$ cat /etc/udev/rules.d/10-wakeup.rules
ACTION=="add", SUBSYSTEM=="usb", ATTRS{idVendor}=="239a", ATTRS{idProduct}=="8120" RUN+="/bin/sh -c 'echo enabled > /sys/bus/usb/devices/usb3/power/wakeup'"
$ cat /sys/bus/usb/devices/usb3/power/wakeup
enabled
```
Unfortunately this doesn't seem to wake up the host. I've confirmed `/sys/bus/usb/devices/usb3/power/wakeup` is enabled each time before and after I test. 

One interesting observation I've made is that when I do put the Linux host to sleep (suspend), the webserver on the Pico is much slower to respond than when I tested it on a Macbook that is in sleep mode. It does eventually give a successful http respond after a couple seconds, verifying the keypress was sent. However it doesn't seem to wake up the computer. I wonder if that has anything to do with it? Perhaps the pico has some kind of low power state that makes it work differently?

**Comment by bendavis78** [comment]
Update: It looks like it was just a matter of getting the devpath correct in the udev rule. I figured out a more rubust way to set the `power/wakeup` attribute (since the devpath can change between boots):

```
ACTION=="add", \
SUBSYSTEM=="usb", \
ATTRS{idVendor}=="239a", ATTRS{idProduct}=="8120", \
ATTR{power/wakeup}="enabled", \
RUN+="/bin/sh -c 'echo set attributes for %p >> /var/log/wakeup-device.log'"
```

The `idVendor` and `idProduct` are what I found in `lsusb`. Not sure if that will be the same for everyone, but that part will be consistent between boots.

**Comment by dhalbert** [comment]
I see it's possible to use wildcards in udev rules: https://unix.stackexchange.com/questions/752425/writing-udev-rules-wildcards, so you could wildcard the `idProduct` if you need to.

**Pull request merged**